data_models/admin.py

```diff
@@ -53,10 +54,37 @@ class Meta:
         fields = '__all__'
 
 
+def get_show_options():
+    shows = rr_api.get_shows()
```

Comment:
rr_api throws an exception when the API is not available. It should still be possible to edit a Show even though the API is down. Should be wrapped with a try/except, and populate the field with the existing value if API is down, or `None` if no value has been set.

Reply:
I think that's impossible. `get_show_options` knows nothing about the object being edited.

Reply:
Here's some thinking aloud, I'm sorry for being so wordy :P

I tried a solution where no options were given if an error occurred while fetching Digas shows (thus only partly implementing your suggestion, specifically the part which is possible). The following happens when the API is inaccessible:
1. The select box is empty, as expected.
2. When the user saves, the show's `digas_id` is then set to `None`.
   * The user is not alerted of this.

The problem is only discovered when the user sees that the show no longer has a link to their podcast in the frontend. They may not notice this before several weeks or months have passed. For old programs, it may never be noticed. It doesn't help that shows are not edited very often, so users are not so familiar with that part of the Django Admin.

I don't think this is acceptable. I think it is more acceptable for the edit form to fail, developers get notified by users, and fixing the problem before it sneakily alters shows. I'm sure you agree.

The approaches we can take, in order of increasing effort:
1. Accept that shows may not be edited while the Radio REST API is inaccessible
2. Extend `TypedChoiceField` to handle the case where the `choices` callable raises an exception, basically implementing your suggestion
3. Create new model for Digas shows (not shown/editable in the Django Admin). Create a `manage.py` command for synchronizing it with the Radio REST API. Run this command regularly, e.g. as a cronjob. Change `digas_id` into a ForeignKey field referencing this new model. This would effectively eliminate the error and make for a more stable and speedy Django Admin when editing shows. 

I think I'll explore the second option, it has some promise.

Reply:
Long story short, this is fixed now by doing some magic when extending the ModelForm for the ShowAdminForm. Basically, it works exactly like you described in your initial comment.

I tried an approach with custom fields, but their code is run when Django is started up. The code therefore does not have access to the digas_id of the show in question.

I also explored an approach with custom widget, which seemed promising. However, when you tried to submit, the field would throw a validation error, since it saw a value which, to it, was not an option. So adding an option in the widget (but not the field) is not an, err, option.

The alternative approaches I explored are available at https://github.com/tobinus/kapina-backend/tree/exploring-digas-id-fallback/django_safe_choice